display_list/display_list_color_source.h

```diff
@@ -636,6 +655,76 @@ class DlSweepGradientColorSource final : public DlGradientColorSourceBase {
   FML_DISALLOW_COPY_ASSIGN_AND_MOVE(DlSweepGradientColorSource);
 };
 
+class DlRuntimeEffectColorSource final : public DlColorSource {
+ public:
+  DlRuntimeEffectColorSource(
+      sk_sp<SkRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      sk_sp<SkData> uniform_data)
+      : runtime_effect_(std::move(runtime_effect)),
+        samplers_(std::move(samplers)),
+        uniform_data_(std::move(uniform_data)) {}
+
+  const DlRuntimeEffectColorSource* asRuntimeEffect() const override {
+    return this;
+  }
+
+  std::shared_ptr<DlColorSource> shared() const override {
+    return std::make_shared<DlRuntimeEffectColorSource>(
+        runtime_effect_, samplers_, uniform_data_);
+  }
+
+  DlColorSourceType type() const override {
+    return DlColorSourceType::kRuntimeEffect;
+  }
+  size_t size() const override { return sizeof(*this); }
+
+  bool is_opaque() const override { return false; }
+
+  sk_sp<SkRuntimeEffect> runtime_effect() const { return runtime_effect_; }
+  std::vector<std::shared_ptr<DlColorSource>> samplers() const {
```

Comment:
All of these should be const return values (immutable)

Reply:
(runtime_effect / samplers / uniform_data)

Reply:
done